crates/nu-cmd-dataframe/src/dataframe/values/nu_dataframe/conversion.rs

```diff
@@ -996,10 +955,252 @@ fn series_to_values(
+            let nanos = match time_unit {
+                TimeUnit::Nanoseconds => *a,
+                TimeUnit::Microseconds => *a * 1_000,
+                TimeUnit::Milliseconds => *a * 1_000_000,
+            };
+            Ok(Value::duration(nanos, span))
+        }
+        // AnyValue::Time represents the current time since midnight.
+        // Unfortunately, there is no timezone related information.
+        // Given this, calculate the current date from UTC and add the time.
+        AnyValue::Time(nanos) => time_from_midnight(*nanos, span),
+        AnyValue::List(series) => {
+            series_to_values(series, None, None, span).map(|values| Value::list(values, span))
+        }
+        AnyValue::Struct(idx, struct_array, s_fields) => {
+            let cols: Vec<String> = s_fields.iter().map(|f| f.name().to_string()).collect();
+            let vals: Result<Vec<Value>, ShellError> = struct_array
+                .values()
+                .iter()
+                .enumerate()
+                .map(|(pos, v)| {
+                    let f = &s_fields[pos];
+                    arr_to_value(&f.dtype, &**v, *idx, span)
+                })
+                .collect();
+            let record = Record { cols, vals: vals? };
+            Ok(Value::record(record, span))
+        }
+        AnyValue::StructOwned(struct_tuple) => {
+            let values: Result<Vec<Value>, ShellError> = struct_tuple
+                .0
+                .iter()
+                .map(|s| any_value_to_value(s, span))
+                .collect();
+            let fields = struct_tuple
+                .1
+                .iter()
+                .map(|f| f.name().to_string())
+                .collect();
+            Ok(Value::Record {
+                val: Record {
+                    cols: fields,
+                    vals: values?,
+                },
+                internal_span: span,
+            })
+        }
+        AnyValue::Utf8Owned(s) => Ok(Value::string(s.to_string(), span)),
+        AnyValue::Binary(bytes) => Ok(Value::binary(*bytes, span)),
+        AnyValue::BinaryOwned(bytes) => Ok(Value::binary(bytes.to_owned(), span)),
+        e => Err(ShellError::GenericError(
+            "Error creating Value".into(),
+            "".to_string(),
+            None,
+            Some(format!("Value not supported in nushell: {e}")),
+            Vec::new(),
+        )),
+    }
+}
+
+#[inline]
+fn arr_to_value(
+    dt: &DataType,
+    arr: &dyn Array,
+    idx: usize,
+    span: Span,
+) -> Result<Value, ShellError> {
+    macro_rules! downcast {
+        ($casttype:ident) => {{
+            let arr = &*(arr as *const dyn Array as *const $casttype);
+            arr.value_unchecked(idx)
+        }};
+    }
+
+    // Not loving the unsafe here, however this largely based off the one
+    // example I found for converting Array values in:
+    // polars_core::chunked_array::ops::any_value::arr_to_any_value
+    unsafe {
+        match dt {
```

Comment:
As the safety condition depends on the correct `DataType` being passed, this function itself has to be `unsafe`.

Even if we declare it as such I really find it hard to validate that this whole operation is sound. 
The call sites of this function need to guarantee both a valid type is ascribed and that all requirements with regards to the pointer and index are met.

Beyond that I am not super comfortable with a manul type reinterpretation that depends on the data representation as given to us by an external crate in this case both `polars_core` and `arrow2`. The former has been quite willing to break APIs so it feels to me like a ticking time bomb to make safety assumptions across crates outside of a simpler well documented interface.
If I have to read the crate source code to reassure me that this is sound and can't just read the documentation of a constrained `unsafe` API I think this is just asking for future trouble.